middle_end/flambda2/from_lambda/lambda_to_flambda_primitives.ml

```diff
@@ -887,27 +1015,87 @@ let array_access_validity_condition array array_kind index
 
 let check_array_access ~dbg ~array (array_kind : P.Array_kind_for_length.t)
     ~index ~index_kind ~size_int primitive : H.expr_primitive =
+  let width_in_scalars =
+    match array_kind with
+    | Array_kind Immediates
+    | Array_kind Values
+    | Array_kind Naked_floats
+    | Array_kind Naked_float32s
+    | Array_kind Naked_int32s
+    | Array_kind Naked_int64s
+    | Array_kind Naked_nativeints
+    | Array_kind Naked_vec128s
+    | Float_array_opt_dynamic ->
+      1
+    | Array_kind (Unboxed_product kinds) -> List.length kinds
```

Comment:
I think this should be the sum of the widths for `kinds`, since they themselves can contain unboxed products.

Reply:
Fixed, and code moved to Flambda_primitive.Array_kind and .Array_kind_for_length